x/gov/types/proposal.go

```diff
@@ -86,6 +90,71 @@ func (p Proposal) UnpackInterfaces(unpacker types.AnyUnpacker) error {
 	return unpacker.UnpackAny(p.Content, &content)
 }
 
+// NewProposalV2 creates a new Proposal2 instance
+func NewProposalV2(
+	messages []sdk.Msg,
+	id uint64,
+	submitTime, depositEndTime time.Time,
+) (ProposalV2, error) {
+
+	msgs := make([]*types.Any, len(messages))
+	for i, msg := range messages {
+		m, ok := msg.(proto.Message)
+		if !ok {
+			return ProposalV2{}, fmt.Errorf("can't proto marshal %T", msg)
+		}
+		any, err := types.NewAnyWithValue(m)
+		if err != nil {
+			return ProposalV2{}, err
+		}
+
+		msgs[i] = any
+	}
+
+	p := ProposalV2{
+		ProposalId:       id,
+		Messages:         msgs,
+		Status:           StatusDepositPeriod,
+		FinalTallyResult: EmptyTallyResult(),
+		TotalDeposit:     sdk.NewCoins(),
+		SubmitTime:       submitTime,
+		DepositEndTime:   depositEndTime,
+	}
+
+	return p, nil
+}
+
+// String implements stringer interface
+func (p ProposalV2) String() string {
+	out, _ := yaml.Marshal(p)
+	return string(out)
+}
+
+func (p ProposalV2) GetMessages() ([]sdk.Msg, error) {
+	msgs := make([]sdk.Msg, len(p.Messages))
+	for i, msgAny := range p.Messages {
+		msg, ok := msgAny.GetCachedValue().(sdk.Msg)
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "messages contains %T which is not a sdk.MsgRequest", msgAny)
+		}
+		msgs[i] = msg
+	}
+
+	return msgs, nil
+}
+
+// UnpackInterfaces implements UnpackInterfacesMessage.UnpackInterfaces
+func (p ProposalV2) UnpackInterfaces(unpacker types.AnyUnpacker) error {
+	var msg sdk.Msg
```

Comment:
you can consider using functions from `types/tx/msgs.go` (see how they are being used for `x/group` proposals).

Reply:
Thanks for the pointer